src/WooCommerceCheckout.php

```diff
@@ -28,6 +28,36 @@ public static function init() {
       add_filter('woocommerce_checkout_fields', __CLASS__ . '::addConfirmationEmailCheckoutField');
       add_action('woocommerce_checkout_process', __CLASS__ . '::checkConfirmationEmailField');
     }
+
+    add_filter( 'woocommerce_form_field', __CLASS__ . '::woocommerceFormField', 10, 4 );
+  }
+
+  /**
+   * Adds inline error messages to checkout fields when needed.
+   *
+   * @implements woocommerce_form_field
+   *
+   * @return string
+   */
+  public static function woocommerceFormField($field, $key, $args, $value) {
+    if (strpos($field, '</label>') === FALSE) {
+      return $field;
+    }
+
+    $error = [];
+    if ($args['required'] ?? FALSE) {
+      $errorMessage = sprintf(__('%s is a required field.', 'woocommerce'), $args['label']);
+      $error[]      = '<span class="error-required" style="display:none">' . $errorMessage . '</span>';
+    }
+    if ($args['validate'] ?? FALSE) {
+      $errorMessage = __('Invalid field', 'woocommerce');
+      $error[]      = '<span class="error-validate" style="display:none">' . $errorMessage . '</span>';
+    }
+
+    if (empty($error)) {
+      return $field;
+    }
+    return substr_replace($field, join('', $error), strpos($field, '</label>'), 0);
```

Comment:
Does it make sense to check if there is an error to result in a single return value?
```suggestion
    if ($error) {
      $field = substr_replace($field, join('', $error), strpos($field, '</label>'), 0);
    }
    return $field;
    
```

Reply:
I guess is just a matter of taste here, but pretty much same thing